### ISSUE TITLE
'Unknown: GraphTile DirectedEdge index out of bounds' error when running bicycle costing

### ISSUE THREAD

**Issue by bryce-sharman** [valhalla/valhalla#4861]
Hi,

I have been running Valhalla using a fairly small (approximately 5x1 km) test OSM network that I created by trimming a larger Ontario network available at GeoFabrik using osmosis. I'm running Valhalla using the gis-ops Docker image.

I came across an error while trying to create travel time matrices using bicycle costing. I'll show this using the /locate API.

First, note that this works for the walk and auto costings. I'm showing the Python code that calls the requests.

**Walk costing**
```python
import requests
r = requests.get(
    "http://localhost:8002/locate", 
    headers={'Content-Type': 'application/json'}, 
    json={
        'locations': [{'lat': 43.61918,'lon': -79.52809}],
        'costing': 'pedestrian'
    }
)
```
This returns the following JSON response:
```
[{'input_lon': -79.52809, 'nodes': [{'lat': 43.619193, 'lon': -79.528141}], 'input_lat': 43.61918, 'edges': [{'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586966}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586966}]}]
```

**Similarly it works for auto costing:**
```python
r = requests.get(
    "http://localhost:8002/locate", 
    headers={'Content-Type': 'application/json'}, 
    json={
        'locations': [{'lat': 43.61918,'lon': -79.52809}],
        'costing': 'auto',
    }
)
```
which returns the following JSON response
```
[{'input_lon': -79.52809, 'nodes': [{'lat': 43.619193, 'lon': -79.528141}], 'input_lat': 43.61918, 'edges': [{'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586965}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 0.0, 'way_id': 968586966}, {'side_of_street': 'neither', 'correlated_lon': -79.528141, 'correlated_lat': 43.619193, 'percent_along': 1.0, 'way_id': 968586966}]}]
```

**The problem comes when I try the exact same call using bicycle costing:**
```python
# bicycle test
r = requests.get(
    "http://localhost:8002/locate", 
    headers={'Content-Type': 'application/json'}, 
    json={
        'locations': [{'lat': 43.61918,'lon': -79.52809}],
        'costing': 'bicycle'
    }
)
```
In this case I get the following error:
```
{'error_code': 199, 'error': 'Unknown: GraphTile DirectedEdge index out of bounds: 769361,3,250 directededgecount= 132', 'status_code': 400, 'status': 'Bad Request'}
```

Note that the only change between these cases is the costing model. Do you know what I need to change to fix this?

Please note that I have attached my Valhalla logs to this issue.

Thanks so much!

Bryce

[valhalla.log](https://github.com/user-attachments/files/16698705/valhalla.log)

**Comment by kevinkreiser** [comment]
can you point us to the exact pbf you used to create the graph. if you can do that we can debug it and find out where our bug is :smile:

**Comment by bryce-sharman** [comment]
Thanks Kevin for looking into this!

Here's the file. Note that I had to zip it for GitHub to accept it for upload.

Bryce

[queensway_test.osm.zip](https://github.com/user-attachments/files/16711289/queensway_test.osm.zip)

**Comment by iiiKoen** [comment]
Are there any updates on this bug?

**Comment by chrstnbwnkl** [comment]
I can't reproduce the described bug using the osm file @bryce-sharman provided and the docker image. This could mean one of these things:

- you configured the docker container with some non-default config options
- there was a bug but it got fixed recently 

To me, this sounds like a wrong tile kind of issue. This can easily happen when working with an edge and its opposing edge, which might be in a different tile. Happy to dig into this if someone provides an example I can reproduce!

**Comment by prozessor13** [comment]
Hi, I have the exact same issue for a lot of bicycle matrix-requests and for:
`locate?json={"locations":[{"lat":50.9420,"lon":6.9581}],"costing":"bicycle"}`

I use the following setup:
 * the current version of [docker-valhalla](https://github.com/nilsnolde/docker-valhalla) (2fed0c5d845ae165734a49b059ce3570da9fddd4)
 * current version of the [NRW](https://download.geofabrik.de/europe/germany/nordrhein-westfalen-latest.osm.pbf) osm-file
 * current version of the [germay](https://download.gtfs.de/germany/free/latest.zip) gtfs-file
 * and the env-variables: `build_transit=True` and `server_threads=5`

Thx in advance!

**Comment by jtrosy** [comment]
👋 Hey I am reviving this as I seem to be experiencing the same issue.

I'll try to include as much details as I can, but let me know if there is anything additional I can provide.


I've identified the root cause which seems to be from transit, which appears to match the original posters error `index out of bounds: 769361,3,250 directededgecount= 132',` that is calling out level 3 from transit as the problem.

TLDR of what I am observing;

1.  The root cause appears to be transit hubs which generates a large number of edges.
2. In converting transit I observe an error like: _[ERROR] NodeInfo: edge count exceeds max: 1301_
3. In building tiles I observer _[ERROR] More than 120 transit edges_

Then when running Valhalla as an API I observe:

Isochrone:  Error 171: No suitable edges near location
Locate: ERROR 199: Unknown: GraphTile DirectedEdge index out of bounds: 792230,3,412 directededgecount= 411

**Note:** This only happens if I run as an API. For whatever reason running the CLI has no issues and the requests for locate & isochrone succeed. This interaction is a bit confusing to me.


My isolated test setup:

**Valhalla**: Used ghcr.io/valhalla/valhalla:latest

**OSM**: I originally saw the issue in Seattle. So for my isolated test I used **washington-latest.osm.pbf**. I just downloaded the latest file yesterday.

**Transit**: I generated a fake transit network and created a hub network with a large amount of stops to mimic what I am observing. Here's a zip of the transit feed I used:

[transit_bug_trigger.zip](https://github.com/user-attachments/files/24244322/transit_bug_trigger.zip)


**Script**: I have a shell script that ran through all the steps for reference. It ingested/converted transit, built tiles, then made HTTP and CLI requests for 4 modes (bicycle, pedestrian, multimodal, auto). 

[demonstrate_bug.sh](https://github.com/user-attachments/files/24244333/demonstrate_bug.sh)


Requests results observed for coordinates: (47.6101, -122.3382)

─────────────────────────────────────────────────────
Testing mode: AUTO
─────────────────────────────────────────────────────
  CLI isochrone (auto): request succeeded
  HTTP locate (auto): request succeeded
  HTTP isochrone (auto): request succeeded

─────────────────────────────────────────────────────
Testing mode: PEDESTRIAN
─────────────────────────────────────────────────────
  CLI isochrone (pedestrian): request succeeded
  HTTP locate (pedestrian): request succeeded
  HTTP isochrone (pedestrian): request succeeded

─────────────────────────────────────────────────────
Testing mode: BICYCLE
─────────────────────────────────────────────────────
  CLI isochrone (bicycle): request succeeded
  HTTP locate (bicycle): 🎯 ERROR 199: Unknown: GraphTile DirectedEdge index out of bounds: 792230,3,412 directededgecount= 411
  HTTP isochrone (bicycle): Error 171: No suitable edges near location

─────────────────────────────────────────────────────
Testing mode: MULTIMODAL
─────────────────────────────────────────────────────
  CLI isochrone (multimodal): request succeeded
  HTTP locate (multimodal): request succeeded
  HTTP isochrone (multimodal): request succeeded




**Logs**

Transit Convert log: 
[transit_convert.log](https://github.com/user-attachments/files/24244374/transit_convert.log)


**Build tiles logs**
[tiles_build_full.log](https://github.com/user-attachments/files/24244379/tiles_build_full.log)

**Comment by chrstnbwnkl** [comment]
Thanks for the detailed write-up! If you have very dense transit data, it can cause a problem since we can only store up to 127 edges per edge – that's limited by the fact that we allow 7 bits on the node to store its outgoing edge count. 

Looking at where that error log statement comes from:
https://github.com/valhalla/valhalla/blob/23d71ae69bda31597a4945e344d1db78ecfb3736/src/mjolnir/transitbuilder.cc#L210-L219


That max count should probably be 127, as the comment above suggests, and that causes a max of 120 edges per node to be added, but on the node we still store a count of 127. 

Fixing that will only prevent the indexing issue of course, not the underlying problem, and I'm not sure there is a good solution to that. I'm not really familiar with GTFS data, how realistic is it to have >127 transit edges running through a node? Seems a bit high at first glance

**Comment by jtrosy** [comment]
> I'm not really familiar with GTFS data, how realistic is it to have >127 transit edges running through a node? Seems a bit high at first glance.

I don't believe it should be very common. In my real life example it was observed in a hub in downtown Seattle where basically all bus routes pass through. Unfortunately this means all bike isochrones in that downtown area fail.

I do think it's strange that the issue surfaces as an issue with bike isochrones - while all other modes are unaffected. Additionally I didn't observe any issues with routing with mode=bicycle. So it seems contained to bike isochrones/locate.

**Comment by chrstnbwnkl** [comment]
Yeah that is strange indeed, I'm not even sure why with bicycle costing we'd even look at transit tiles in the first place... I will try to find some time to look at it a bit closer.

**Comment by chrstnbwnkl** [comment]
I've had a closer look, and indeed the error disappears once the max transit edges is set to 127. 

> I do think it's strange that the issue surfaces as an issue with bike isochrones - while all other modes are unaffected.  

According to your logs the error doesn't surface on isochrones though, only on `/locate`. Not sure why bicycle is the only mode triggering the error, but I wouldn't focus too much on that. As far as I remember, valhalla will load transit level tiles even in non-transit costing, and discard them based on their use. In this case it won't even get there, because it looks like we store an index to a transit edge that we deleted from the final graph tile :smile:

The bad news is my comment above still holds: we can prevent the error happening at run-time, but we can't easily fix the underlying problem, which is that we can only add a limited number of edges per node, which may be too low in dense transit networks.

**Comment by jtrosy** [comment]
> we can prevent the error happening at run-time, but we can't easily fix the underlying problem, which is that we can only add a limited number of edges per node, which may be too low in dense transit networks.

Yeah there are definitely layers to this issue.

The out of index being fixed is appreciated for sure though.

**Comment by chrstnbwnkl** [comment]
Is the GTFS data you originally observed this on open data? Would be interesting to take a look. It's possible that the producer of the feed wrongly encodes variations to a route as distinct routes which ends up causing a lot of edges to get created by Valhalla.

**Comment by jtrosy** [comment]
> Is the GTFS data you originally observed this on open data? 

It is not sorry.

I can take a deeper dive into the GTFS feed I have; but I am not sure I'd catch an issue with the feed unless it was super obvious.
